Reporter vector row sum and dot product

lynnmunday opened the pull request:
closes #26198

lynnmunday committed:
initial working version of reporter that sums vector of vectors into a single vector

lynnmunday committed:
Adding documentation, testing and renaming object to be more descriptive.  This completes the implementation for the vectorOfVectorRowSum ref #26198

lynnmunday committed:
reporter for vector dot product closes #26198

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1912213/) on 340ff87 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/26202/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1912254/) on 340ff87 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>6a8607</th>
      <th colspan=4>#26202 340ff8</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.66%</td>
      <td>85.66%</td>
      <td>+0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>93639</td>
      <td>93640</td>
      <td>+1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15676</td>
      <td>15675</td>
      <td>-1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26202/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26202/coverage/framework)


## Modules coverage

### Optimization

<table>
  <thead>
    <tr>
      <th></th>
      <th>6a8607</th>
      <th colspan=4>#26202 340ff8</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.47%</td>
      <td>89.76%</td>
      <td>+0.28%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>1751</td>
      <td>1805</td>
      <td>+54</td>
      <td>54</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>206</td>
      <td>206</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26202/coverage_diff/optimization)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26202/coverage/optimization)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/26202/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/26202/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/26202/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/26202/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/26202/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/26202/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/26202/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/26202/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/26202/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/26202/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/26202/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/26202/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/26202/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/26202/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/26202/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/26202/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/26202/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/26202/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/26202/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/26202/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/26202/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/26202/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/26202/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/26202/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/26202/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/26202/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/26202/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/26202/coverage/xfem)

</details>



This comment will be updated on new commits.

lynnmunday committed:
moved test to organize optimization tests

lynnmunday commented:
This is wip.  I need to add more reporters and examples

GiudGiud commented on modules/optimization/doc/content/source/reporters/VectorDotProduct.md:
```diff
@@ -0,0 +1,13 @@
+# VectorDotProduct
```
actually StdVectorRealReporterParsedOperation and you pass a parsed expression that is applied to every component

lynnmunday replied:
this is good idea, I have to think about it. I could easily do the stuff in https://mooseframework.inl.gov/source/postprocessors/VectorPostprocessorReductionValue.html
but I don't think this is how you'd want to generalize reporter operations that contain Real values.   I just realized I need to add another reporter that will sum two scalars.  So now there is a row sum for a vector of vectors, vector dot product, vector sum into a scalar and scalar sum.

GiudGiud replied:
@loganharbour do you think we handle the type of the reporter dynamically?
As in we pass
"input_reporter_type = Vector"
"input_data_type = Real"
"output_reporter_type = Scalar"
"output_data_type = Real"

loganharbour replied:
I'm pretty sure that we'll be required to make the type an input option

lynnmunday replied:
I agree with the usefulness of @GiudGiud and @loganharbour comments about a more general reporter with operations object but I'm not the right person to develop something like that.  In the mean time, I can get these reporters and PR merged and replace them later when a better option becomes available. 

These reporters that operate on their data aren't useful for standard simulations because they are just doing a post-processing step that could easily be done in python since all of the reporter data is going to be written to json anyway.  So maybe it is not worth the effort to develop a more general reporter with operations object.  For optimization, I need to process the data and then make a decision using that data within the same simulation, so this can't be done as a postprocessing step.

GiudGiud replied:
There's plenty of examples of parsed operation objects at this point, it's not as complicated as it used to be. 

I dont think it's sustainable to implement an object for basic operations, and while this was the initial approach for auxkernels and PPs, it was just not the right approach. We cant solve a combinatorial problem (N operations * M input data types * M output data types * L systems (auxkernels, pps, reporters, vpps...)) by addressing every combination individually. I think doing parsing + templating would get us though the N*M*M part of the problem.

lynnmunday committed:
adding reporter to sum components of a vector

lynnmunday committed:
combined example referenced from optmiization module showing optmization of multiple forward problems using the sampler system

lynnmunday committed:
Changed an inverse optimization problem to use generalOptimization and computing
the objective value directly on the forwardAndAdjoint problem using
VectorDotProduct.  This file also shows how weights can be used with the
GeneralOptimization reporter where the weights must be read in with the
measurement values into the OptimizationData reproter on the sub-app.  Using weights
gives the same optmized parameters and temperature field but changes the adjoint
temperature field because the dirac Kernels are scaled, which led to the
exodus file being re-golded.
closes #26198

lynnmunday committed:
fixing broken documentation closes #26198

lynnmunday committed:
generalize vector dot product to operate on lots of vectors with a parsed function

lynnmunday committed:
changing vectorSum to take a parsed function to do the reduction

lynnmunday committed:
function to row sum vector of vectors and get tests to pass.

lynnmunday committed:
adding parsed function for scalar reporters and changing name of similar vector reporter object
fixed tests to use new object names

lynnmunday committed:
updated documentation

lynnmunday commented on modules/optimization/doc/content/source/reporters/VectorSum.md:
```diff
@@ -0,0 +1,30 @@
+
+## Overview
+
+`VectorSum` operates on the elements of a vector in a reporter using a parsed function containing only two variables, `vi` and `vplus`.
```
@giud @dschwen Would you recommend some different wording for this description and comment on the implementation of this object?  It is doing the vector reduction we talked about using a parsed function to define the function that will be used to accumulate the result as it iterates through the components of a vector.  It is doing what I want but I needs better naming.  The functions are also kind of awkward.  The class name is also bad.

GiudGiud replied:
ParsedVectorReduction ? Or ParsedRealStdVectorReduction ?
it s going from std::vector to 1 Real right?

lynnmunday replied:
Yes, this one goes from a vector to a scalar and the vectorOfvector goes from a matrix to a vector.

GiudGiud replied:
ParsedStdVectorRealReductionReporter would work?
Parsed -> parsed expression
Reduction -> for the dimensionality reduction
StdVectorReal -> only type it supports?
Reporter -> needed because the same operation on an array auxvariable to a regular field one would be (roughly) the same name otherwise

lynnmunday commented on modules/optimization/src/reporters/ParsedScalars.C:
```diff
@@ -0,0 +1,129 @@
+  const std::vector<std::string> reporter_symbols(
+      getParam<std::vector<std::string>>("reporter_symbols"));
+
+  if (reporter_names.size() != reporter_symbols.size())
```
I'll add a base class for ParsedScalars and ParsedVectors to sets up the common part in the constructor.

GiudGiud replied:
You should append Reporter to the name so there's no overlap

lynnmunday replied:
@GiudGiud Will you take another look at the code, input files and naming.  I wanted to get that all correct before I fixed the documentation.

dschwen reviewed (commented):
At first glance I see a ton of duplicated code.

lynnmunday committed:
derived from common parsed function reporter base class and updated documentation

lynnmunday commented:
> At first glance I see a ton of duplicated code.
@dschwen I fixed the duplication.  Can you look at the code?  The documentation needs udpated.

GiudGiud commented on modules/optimization/doc/content/source/reporters/ParsedScalars.md:
```diff
@@ -0,0 +1,21 @@
+# ParsedScalars
+
+!syntax description /Reporters/ParsedScalars
```
```suggestion
!syntax description /Reporters/ParsedScalarReporter
```

GiudGiud replied:
I would not say Scalars since you only define one scalar reporter right?
then add Reporter at the end for disambiguation

GiudGiud commented on modules/optimization/doc/content/source/reporters/ParsedVectors.md:
```diff
@@ -0,0 +1,17 @@
+# ParsedVectors
+
+!syntax description /Reporters/ParsedVectors
```
```suggestion
!syntax description /Reporters/ParsedVectorReporter
```

GiudGiud commented on modules/optimization/src/reporters/ParsedReporterBase.C:
```diff
@@ -0,0 +1,88 @@
+  params.addRequiredCustomTypeParam<std::string>(
+      "expression", "FunctionExpression", "function expression");
+  params.addParam<std::string>("name", "result", "Name of output reporter.");
+  params.addRequiredParam<std::vector<std::string>>("reporter_symbols",
```
can these be ReporterName or are you getting errors from lack of /

GiudGiud commented on modules/optimization/src/reporters/ParsedReporterBase.C:
```diff
@@ -0,0 +1,88 @@
+{
+  // build reporters argument
+  std::string symbol_str;
+  for (std::size_t i = 0; i < _reporter_symbols.size(); ++i)
```
```suggestion
  for (const auto i : index_range(_reporter_symbols))
```

dschwen replied:
There are many more places where range based loops could be used

dschwen replied:
Well... five to be precise

lynnmunday replied:
I added them.

lynnmunday committed:
addressing reviewer comments closes #26198

lynnmunday commented:
@GiudGiud I've addressed your review comments.  Could I get this merged and if there are additional changes, they can be addressed in a new pr.

GiudGiud commented:
Was waiting on @dschwen's opinion on these parsed matters

lynnmunday commented:
@GiudGiud I've made all of the requested changes to this PR and it is going in the optimization module.  This PR shouldn't be delayed any longer by additional feature requests that don't address the original issue, those should be addressed by a new PR.

lynnmunday commented:
@dschwen will you review and merge?

dschwen reviewed (approved):
Missing range based loops should not hold this up.

lynnmunday commented:
sorry @dschwen.  I had an extra file in the commit that I needed to drop.

lynnmunday committed:
use make_range closes #26198

GiudGiud commented:
@loganharbour

loganharbour commented on modules/optimization/include/reporters/ParsedScalarReporter.h:
```diff
@@ -0,0 +1,33 @@
+  virtual void finalize() override;
+
+private:
+  /// output vector
```
```suggestion
  /// output scalar
```

loganharbour commented on modules/optimization/include/reporters/ParsedScalarReporter.h:
```diff
@@ -0,0 +1,33 @@
+  /// output vector
+  Real & _output_reporter;
+
+  std::vector<const Real *> _reporter_data;
```
```suggestion
  /// Input reporter values
  std::vector<const Real *> _reporter_data;
```

loganharbour commented on modules/optimization/include/reporters/ParsedVectorReporter.h:
```diff
@@ -0,0 +1,31 @@
+  virtual void finalize() override;
+
+private:
+  std::vector<const std::vector<Real> *> _reporter_data;
```
```suggestion
  /// input reporter vectors
  std::vector<const std::vector<Real> *> _reporter_data;
```

loganharbour commented on modules/optimization/src/reporters/ParsedReporterBase.C:
```diff
@@ -0,0 +1,88 @@
+
+    // wait for ranks > 0 to catch up
+    if (_communicator.rank() == 0)
+      _communicator.barrier();
```
@dschwen is this how we do this elsewhere?

lynnmunday replied:
Most of this constructor is copied from ParsedAux.

loganharbour replied:
👍🏼

loganharbour commented on modules/optimization/src/reporters/ParsedScalarReporter.C:
```diff
@@ -0,0 +1,54 @@
+  const std::vector<ReporterName> reporter_names(
+      getParam<std::vector<ReporterName>>("reporter_names"));
+  if (reporter_names.size() != _reporter_symbols.size())
+    mooseError(
```
Should be a param error on one of the parameters and avoid using newlines in parameter errors

lynnmunday replied:
The new lines make it easier to read.  Do you really want them gone?

loganharbour commented on modules/optimization/src/reporters/ParsedScalarReporter.C:
```diff
@@ -0,0 +1,54 @@
+ParsedScalarReporter::validParams()
+{
+  InputParameters params = ParsedReporterBase::validParams();
+  params.addRequiredParam<std::vector<ReporterName>>("reporter_names", "Reporter names ");
```
```suggestion
  params.addRequiredParam<std::vector<ReporterName>>("reporter_names", "The input reporter names");
```

loganharbour commented on modules/optimization/src/reporters/ParsedScalarReporter.C:
```diff
@@ -0,0 +1,54 @@
+{
+  InputParameters params = ParsedReporterBase::validParams();
+  params.addRequiredParam<std::vector<ReporterName>>("reporter_names", "Reporter names ");
+  params.addClassDescription("Apply parsed functions to scalar entries held in reporters.");
```
```suggestion
  params.addClassDescription("Applies parsed functions to scalar entries held in reporters.");
```

loganharbour commented on modules/optimization/src/reporters/ParsedVectorReporter.C:
```diff
@@ -0,0 +1,76 @@
+  const std::vector<ReporterName> reporter_names(
+      getParam<std::vector<ReporterName>>("reporter_names"));
+  if (reporter_names.size() != _reporter_symbols.size())
+    mooseError(
```
param error

loganharbour commented on modules/optimization/test/tests/reporters/vector_math/tests:
```diff
@@ -0,0 +1,28 @@
+    allow_test_objects = true
+    requirement = "The system shall be able to row sum a vector of vectors into a single vector, perform a dot product between two vectors and sum all of the entries of a single vector."
+  []
+  [errorRowSum]
+    type = RunException
+    input = vectorMath.i
+    cli_args = "Reporters/dataFromVofV/vector_of_vectors='101 201; 102 202; 103'"
+    allow_test_objects = true
+    expect_err = "Every vector in 'reporter_name=dataFromVofV/v_of_v' must be the same size."
+    requirement = "The system shall throw an error when the vectors being reduced are not the same size."
+  []
+  [errorDotProduct]
+    type = RunException
+    input = vectorMath.i
+    cli_args = "Reporters/vectorOperation/reporter_names='vecs/vec_a vecs/vec_b vec_d/vec_d'"
+    allow_test_objects = true
+    expect_err = "All vectors being operated on must be the same size."
+    requirement = "The system shall throw an error when the vectors being operated on are not the same size."
+  []
```
These requirements need a little more context, something like:

- The system shall throw a reasonable error when reducing vectors and the vectors into a single vector and the input vectors are not the same size
- The system shall report a reasonable error when ...

loganharbour reviewed (changes requested):
This all looks reasonable to me. Moderate requests

lynnmunday committed:
addressing review comments. closes #26198

lynnmunday commented:
Thanks @loganharbour   I addressed your comments.

lynnmunday commented:
@loganharbour will you take another look.  The failures are not related to my code.

loganharbour reviewed (approved):
Failures are unrelated.
